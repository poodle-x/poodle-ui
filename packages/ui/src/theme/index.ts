import React from "react";
import {
	ThemeValue,
	SystemThemeConfig,
	BreakpointsThemeConfig,
	SpacingThemeConfig,
	ColorsThemeConfig,
	FontsThemeConfig,
	BordersThemeConfig,
	ShadowThemeConfig,
	LayoutsThemeConfig,
} from "@poodle/system";

import { CSSObjectSystem, Stylesheet } from "../styled";
import { BoxThemeConfig } from "../Box";
import { ButtonThemeConfig } from "../Button";
import { ColumnsThemeConfig } from "../Columns";
import { ColumnThemeConfig } from "../Column";
import { ContainerThemeConfig } from "../Container";
import { TableThemeConfig } from "../Table";
import { TableHeadThemeConfig } from "../TableHead";
import { TableBodyThemeConfig } from "../TableBody";
import { TableRowThemeConfig } from "../TableRow";
import { TableCellThemeConfig } from "../TableCell";
import { VisuallyHiddenThemeConfig } from "../VisuallyHidden";
import { CheckControlThemeConfig } from "../CheckControl";
import { CheckboxThemeConfig } from "../Checkbox";
import { RadioThemeConfig } from "../Radio";
import { InputThemeConfig } from "../Input";
import { InputAdornmentThemeConfig } from "../InputAdornment";
import { IconButtonThemeConfig } from "../IconButton";
import { ModalThemeConfig } from "../Modal";
import { ModalHeaderThemeConfig } from "../ModalHeader";
import { ModalBodyThemeConfig } from "../ModalBody";
import { ModalFooterThemeConfig } from "../ModalFooter";
import { SelectThemeConfig } from "../Select";
import { TabsThemeConfig } from "../Tabs";
import { TabThemeConfig } from "../Tab";
import { TabListThemeConfig } from "../TabList";
import { TabPanelThemeConfig } from "../TabPanel";

export { Stylesheet };

export { useTheme } from "./useTheme";

export type DirectionThemeConfig = "rtl" | "ltr";

export interface GlobalThemeConfig {
	styles?: ThemeValue<CSSObjectSystem>;
}

export type IconType = React.FC<
	React.PropsWithoutRef<React.SVGProps<SVGSVGElement>> &
		React.RefAttributes<SVGSVGElement>
>;

export interface BaseIcons {
	RadioButton?: IconType;
	RadioCheckedButton?: IconType;
	CheckboxButton?: IconType;
	CheckboxCheckedButton?: IconType;
	CheckboxIndeterminateButton?: IconType;
	ArrowDown?: IconType;
}

export interface ThemeConfig extends SystemThemeConfig {
	global?: GlobalThemeConfig;
	direction?: ThemeValue<DirectionThemeConfig>;
	baseIcons?: BaseIcons;
	Box?: BoxThemeConfig;
	Button?: ButtonThemeConfig;
	Columns?: ColumnsThemeConfig;
	Column?: ColumnThemeConfig;
	Container?: ContainerThemeConfig;
	Table?: TableThemeConfig;
	TableHead?: TableHeadThemeConfig;
	TableBody?: TableBodyThemeConfig;
	TableRow?: TableRowThemeConfig;
	TableCell?: TableCellThemeConfig;
	VisuallyHidden?: VisuallyHiddenThemeConfig;
	CheckControl?: CheckControlThemeConfig;
	Checkbox?: CheckboxThemeConfig;
	Radio?: RadioThemeConfig;
	Input?: InputThemeConfig;
	InputAdornment?: InputAdornmentThemeConfig;
	IconButton?: IconButtonThemeConfig;
	Modal?: ModalThemeConfig;
	ModalHeader?: ModalHeaderThemeConfig;
	ModalBody?: ModalBodyThemeConfig;
	ModalFooter?: ModalFooterThemeConfig;
	Select?: SelectThemeConfig;
	Tabs?: TabsThemeConfig;
	Tab?: TabThemeConfig;
	TabList?: TabListThemeConfig;
	TabPanel?: TabPanelThemeConfig;
}

export const defaultGlobal = (
	overrideTheme: ThemeConfig
): GlobalThemeConfig => {
	return {
		styles: overrideTheme.global?.styles,
	};
};

export const defaultBreakpoints = (
	overrideTheme: ThemeConfig
): BreakpointsThemeConfig => {
	return {
		scale: ["sm", "md", "lg", "xl"],
		...overrideTheme.breakpoints,
		sets: {
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
			xxl: 1400,
			...overrideTheme.breakpoints?.sets,
		},
	};
};

export const defaultSpacing = (
	overrideTheme: ThemeConfig
): SpacingThemeConfig => {
	return {
		scale: 4,
		...overrideTheme.spacing,
		sets: {
			s: 4,
			m: 8,
			l: 16,
			...overrideTheme.spacing?.sets,
		},
	};
};

export const defaultFonts = (overrideTheme: ThemeConfig): FontsThemeConfig => {
	return {
		scale: "1rem",
		...overrideTheme.fonts,
		base: {
			fontFamily:
				'-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
			fontStyle: "normal",
			fontWeight: "normal",
			lineHeight: "125%",
			fontSize: "1rem",
			...overrideTheme.fonts?.base,
		},
		weights: {
			normal: 400,
			semibold: 600,
			bold: 700,
			...overrideTheme.fonts?.weights,
		},
		sets: {
			input: {
				fontSize: "1rem",
				fontWeight: "normal",
				lineHeight: "125%",
			},
			button: {
				fontSize: "1rem",
				fontWeight: "bold",
				lineHeight: "150%",
			},
			label: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			helper: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelMedium: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1rem",
				lineHeight: "110%",
			},
			checkControlLabelSmall: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelLarge: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1.125rem",
				lineHeight: "110%",
			},
			modalTitle: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "1.5rem",
				lineHeight: "133.33%",
			},
			...overrideTheme.fonts?.sets,
		},
	};
};

export const defaultColor = (overrideTheme: ThemeConfig): ColorsThemeConfig => {
	return {
		modes: {
			...overrideTheme.colors?.modes,
		},
		sets: {
			text: "#282828",
			disabledText: "#A6A6A6",

			primary: {
				light: "#D1EBFF",
				base: "#00478A",
				dark: "#003A70",
				textOnLight: "#00478A",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			negative: {
				base: "#AF1308",
				dark: "#8D1007",
				light: "#FCF3F3",
				textOnLight: "#AF1308",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			positive: {
				base: "#006122",
				dark: "#004D1B",
				light: "#E5FFEE",
				textOnLight: "#006122",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			warn: {
				base: "#FFBB33",
				dark: "#FAAB00",
				light: "#FFF1DB",
				textOnLight: "#282828",
				textOnBase: "#282828",
				textOnDark: "#282828",
			},

			bg: {
				base: "#F8F8F8",
				100: "#E0E0E0",
				textOn100: "#282828",
				200: "#CCCCCC",
				textOn200: "#282828",
			},

			modal: {
				base: "#F8F8F8",
				footer: "#F0F0F0",
			},

			border: {
				base: "#CCCCCC",
			},

			...overrideTheme.colors?.sets,
		},
	};
};

export const defaultBorder = (
	overrideTheme: ThemeConfig
): BordersThemeConfig => {
	return {
		sets: {
			base: {
				style: "solid",
				width: "1px",
				color: "border",
			},
			...overrideTheme?.borders?.sets,
		},
		radius: {
			base: "4px",
			...overrideTheme?.borders?.radius,
		},
	};
};

export const defaultBoxShadow = (
	overrideTheme: ThemeConfig
): ShadowThemeConfig => {
	return {
		boxShadowSets: {
			lv1: "1px 2px 2px -1px rgba(0,0,0,0.4)",
			lv2: "2px 2px 5px 0px rgba(0,0,0,0.4)",
			lv3: "4px 4px 5px 0px rgba(0,0,0,0.4)",
			...overrideTheme?.shadows?.boxShadowSets,
		},
	};
};

export const defaultLayouts = (
	overrideTheme: ThemeConfig
): LayoutsThemeConfig => {
	return {
		zIndices: {
			modal: 2000,
			...overrideTheme?.layouts?.zIndices,
		},
	};
};

export function createTheme(overrideTheme: ThemeConfig = {}): ThemeConfig {
	return {
		direction: "ltr",
		...overrideTheme,
		global: defaultGlobal(overrideTheme),
		spacing: defaultSpacing(overrideTheme),
		breakpoints: defaultBreakpoints(overrideTheme),
		colors: defaultColor(overrideTheme),
		fonts: defaultFonts(overrideTheme),
		borders: defaultBorder(overrideTheme),
		shadows: defaultBoxShadow(overrideTheme),
		layouts: defaultLayouts(overrideTheme),
	};
}

export type StylesConfig<T extends string> = {
	[K in T]: ThemeValue<CSSObjectSystem>;
};

export interface StandardComponentProps {
	variant?: string;
	themeExtend?: ThemeConfig;
	theme?: ThemeConfig;
	sizeStyle?: string;
	colorStyle?: string;
}

export interface StandardThemeConfig<
	T extends string,
	Props extends { [key: string]: any }
> {
	defaultProps?: Partial<Props>;
	styles?: Partial<StylesConfig<T>>;
	variants?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	sizes?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	colorStyles?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	modes?: {
		[key: string]: {
			styles?: Partial<StylesConfig<T>>;
			variants?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			sizes?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			colorStyles?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
		};
	};
	overrides?: Partial<StylesConfig<T>>;
	overrideClasses?: Partial<
		{
			[K in T]: boolean | string;
		}
	>;
}

export * from "./system";
