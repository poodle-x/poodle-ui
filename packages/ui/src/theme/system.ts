import {
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	compose,
	createSystem,
	getMode,
	FontsProps,
	ColorsProps,
	SpacingProps,
	LayoutProps,
	BordersProps,
	bordersSystem,
	flexSystem,
	FlexProps,
	pseudosSystem,
	PseudosProps,
	shadowSystem,
	ShadowProps,
	GridProps,
	gridSystem,
	ListProps,
	listSystem,
	SystemConfigList,
	ResponsiveProp,
} from "@poodle/system";
import { CSSSystem } from "../styled";
import { StandardThemeConfig } from "./index";
import * as CSS from "csstype";

const additionalSystem: SystemConfigList = {
	objectFit: true,
	objectPosition: true,
	transition: true,
	cursor: true,
	pointerEvents: true,
};

export interface AdditionalSystemProps {
	/**
	 * The CSS `object-fit` property
	 */
	objectFit?: ResponsiveProp<CSS.Property.ObjectFit>;
	/**
	 * The CSS `object-fit` property
	 */
	objectPosition?: ResponsiveProp<CSS.Property.ObjectPosition>;
	/**
	 * The CSS `transition` property
	 */
	transition?: ResponsiveProp<CSS.Property.Transition>;
	/**
	 * The CSS `cursor` property
	 */
	cursor?: ResponsiveProp<CSS.Property.Cursor>;
	/**
	 * The CSS `cursor` property
	 */
	pointerEvents?: ResponsiveProp<CSS.Property.PointerEvents>;
}

const allSystem = compose(
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	bordersSystem,
	flexSystem,
	pseudosSystem,
	shadowSystem,
	gridSystem,
	listSystem,
	additionalSystem
);

export const system = createSystem(allSystem);

interface CssSystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		GridProps,
		ListProps {}

export interface SystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		PseudosProps<CssSystemProps>,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

const additionalProps = [
	"textStyle",
	"variant",
	"themeExtend",
	"theme",
	"sizeStyle",
	"colorStyle",
];

export function removeSystemProps(props: { [key: string]: any }) {
	let resultProps = {};
	Object.keys(props).forEach((prop) => {
		let found = false;
		for (const name in allSystem) {
			if (prop === name || additionalProps.indexOf(prop) >= 0) {
				found = true;
				break;
			}
		}

		if (!found && ["sx"].indexOf(prop) === -1) {
			resultProps = {
				...resultProps,
				[prop]: props[prop],
			};
		}
	});

	return resultProps;
}

export function createCSSSystemStandard<Key extends string>(data: {
	key: Key;
	config?: StandardThemeConfig<Key, any>;
	props: { [key: string]: any };
	base?: CSSSystem["base"];
}): CSSSystem {
	const { key, props, config, base } = data;

	const { variant = "", sizeStyle = "", colorStyle = "" } = props;

	const mode = getMode(props);

	const overrideClassName = config?.overrideClasses?.[key];

	return {
		base,
		overrideClassName,
		applies: [
			config?.styles?.[key],
			config?.modes?.[mode]?.styles?.[key],
			config?.variants?.[variant]?.[key],
			config?.modes?.[mode]?.variants?.[variant]?.[key],
			config?.sizes?.[sizeStyle]?.[key],
			config?.modes?.[mode]?.sizes?.[sizeStyle]?.[key],
			config?.colorStyles?.[colorStyle]?.[key],
			config?.modes?.[mode]?.colorStyles?.[colorStyle]?.[key],
		],
		override: config?.overrides?.[key],
	};
}

export * from "@poodle/system";
