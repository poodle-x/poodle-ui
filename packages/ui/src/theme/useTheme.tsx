import { ThemeContext } from "@emotion/react";
import deepmerge from "deepmerge";
import React from "react";
import { ThemeConfig } from "./";

export function useTheme(
	options: {
		override?: ThemeConfig;
		extend?: ThemeConfig;
	} = {}
): { theme: ThemeConfig } {
	const { override, extend } = options;

	const themeContext: ThemeConfig = React.useContext(ThemeContext);

	const returnTheme = React.useMemo(() => {
		if (override) {
			return override;
		}

		if (extend) {
			return deepmerge(themeContext, extend, {
				arrayMerge: (_, sourceArray) => {
					return sourceArray;
				},
			});
		}

		return themeContext;
	}, [extend, override, themeContext]);

	return {
		theme: returnTheme || {},
	};
}
