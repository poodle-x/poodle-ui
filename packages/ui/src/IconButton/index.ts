import { IconButton, IconButtonProps } from "./IconButton";
import { StandardThemeConfig } from "../theme";

export * from "./IconButton";

export default IconButton;

export type IconButtonStyleKeys = "root";

export interface IconButtonThemeConfig
	extends StandardThemeConfig<IconButtonStyleKeys, IconButtonProps> {}
