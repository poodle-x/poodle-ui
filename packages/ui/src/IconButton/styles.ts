import { createCSSSystemStandard } from "../theme";
import { CSSObjectSystem, CSSSystem } from "../styled";
import { IconButtonProps } from "./IconButton";
import { IconButtonStyleKeys } from "./";

export function Root(props: IconButtonProps): CSSSystem {
	const { theme, shapeStyle, disabled, colorStyle } = props;

	let shapeStyles: CSSObjectSystem = {};
	let colorStyles: CSSObjectSystem = {};

	if (!disabled) {
		switch (colorStyle) {
			case "primary": {
				colorStyles = {
					color: "primary",

					"&:hover, &:focus": {
						color: "primary.textOnLight",
						backgroundColor: "primary.light",
					},
				};
				break;
			}
			case "negative": {
				colorStyles = {
					color: "negative",

					"&:hover, &:focus": {
						color: "negative.textOnLight",
						backgroundColor: "negative.light",
					},
				};
				break;
			}
			case "positive": {
				colorStyles = {
					color: "positive",

					"&:hover, &:focus": {
						color: "positive.textOnLight",
						backgroundColor: "positive.light",
					},
				};
				break;
			}
			case "warn": {
				colorStyles = {
					color: "warn",

					"&:hover, &:focus": {
						color: "warn.textOnLight",
						backgroundColor: "warn.light",
					},
				};
				break;
			}
		}
	}

	if (shapeStyle === "round") {
		shapeStyles = {
			borderRadius: "50%",
		};
	}

	if (shapeStyle === "square") {
		shapeStyles = {
			borderRadius: "base",
		};
	}

	return createCSSSystemStandard<IconButtonStyleKeys>({
		key: "root",
		config: theme?.IconButton,
		props,
		base: [
			{
				display: "inline-flex",
				alignSelf: "center",
				justifyContent: "center",
				border: 0,
				padding: "scale-2",
				pointerEvents: "initial",
				fontSize: "1.5rem",
				color: "text",
				cursor: "pointer",
				backgroundColor: "transparent",
				outline: "none",
				appearance: "none",
				transition: "background-color 220ms ease-in, color 220ms ease-in",
				...shapeStyles,

				"&:hover, &:focus": {
					color: "bg.textOn100",
					backgroundColor: "bg.100",
				},

				"&:disabled": {
					color: "disabledText",
					cursor: "not-allowed",
				},
			},
			colorStyles,
		],
	});
}
