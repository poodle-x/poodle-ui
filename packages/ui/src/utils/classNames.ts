import React from "react";
import { css, CSSObjectSystem, cx } from "../styled";
import { ThemeValue } from "../theme";
import { cssSystem } from "./cssSystem";

type ClassName =
	| undefined
	| string
	| ((props: any) => string)
	| { [key: string]: boolean };

export function genClassNames(data: {
	props: any;
	classNames: Array<ClassName>;
}) {
	const { props, classNames } = data;

	let resultClassNames = "";

	classNames.forEach((cn) => {
		if (typeof cn === "function") {
			resultClassNames = cx(resultClassNames, cn(props));
			return;
		}

		resultClassNames = cx(resultClassNames, cn);
	});

	return resultClassNames;
}

export function classNameCompose(data: {
	props: any;
	overrideClassName?: string | boolean | undefined;
	styles: {
		base: CSSObjectSystem[];
		applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
		override?: ThemeValue<CSSObjectSystem>;
	};
}) {
	const { overrideClassName, styles, props } = data;

	if (typeof overrideClassName === "string") {
		return overrideClassName;
	}

	if (overrideClassName === true) {
		return "";
	}

	return css(
		cssSystem({
			...styles,
			props,
		})
	);
}

export function genListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
): { [key in keyof typeof lists]: string } {
	const result: { [key: string]: string } = {};
	Object.keys(lists).forEach((key) => {
		let resultClassNames = "";
		const listKeyData = lists[key];
		resultClassNames = genClassNames({
			props: { ...baseProps, ...listKeyData.props },
			classNames: listKeyData.classNames,
		});
		result[key] = resultClassNames;
	});

	return result;
}

export function useGenListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
) {
	return React.useMemo(() => {
		return genListClassNames(baseProps, lists);
	}, [lists, baseProps]);
}
