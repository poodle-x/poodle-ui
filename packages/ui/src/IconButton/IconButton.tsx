import React from "react";
import Box, { BoxProps } from "../Box";
import { StandardComponentProps, ThemeConfig } from "../theme";
import useDefaultProps from "../utils/useDefaultProps";
import { useClassNames } from "../styled";
import * as styles from "./styles";

export interface LocalIconButtonProps {
	children?: React.ReactNode;
	/**
	 * Change component shape.
	 *
	 * Square shape is useful to create button group.
	 */
	shapeStyle?: "round" | "square";
}

export interface IconButtonProps
	extends BoxProps,
		StandardComponentProps,
		LocalIconButtonProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.IconButton?.defaultProps;
}

export const IconButton: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<IconButtonProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, IconButtonProps>((_props, ref) => {
	const props = useDefaultProps<IconButtonProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});

	const {
		children,
		className,
		href,
		disabled,
		shapeStyle,
		...otherProps
	} = props;

	const classes = useClassNames({
		props,
		lists: {
			root: {
				classNames: [styles.Root, className],
			},
		},
	});

	return (
		<Box
			as={href && !disabled ? "a" : "button"}
			disabled={disabled}
			{...otherProps}
			className={classes.root}
			ref={ref}
		>
			{children}
		</Box>
	);
});

IconButton.displayName = "PoodleIconButton";

IconButton.defaultProps = {
	shapeStyle: "round",
};
