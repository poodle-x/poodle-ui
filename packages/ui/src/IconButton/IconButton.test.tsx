import React from "react";
import { render } from "../utils/test";
import IconButton from "./";
import { ThemeConfig } from "../theme";

describe("<IconButton />", () => {
	it("should render IconButton correctly", () => {
		const { container } = render(<IconButton>IconButton</IconButton>);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render shape IconButton correctly", () => {
		const { container } = render(
			<IconButton shapeStyle="square">IconButton</IconButton>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("ref should work", () => {
		const ref = React.createRef<HTMLDivElement>();

		render(<IconButton ref={ref}>IconButton</IconButton>);
		expect(ref.current).toMatchSnapshot();
	});

	it("should render Box props correctly", () => {
		const { container } = render(
			<IconButton id="id" mt="20px">
				IconButton
			</IconButton>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	describe("theming", () => {
		const themeTest: ThemeConfig = {
			mode: "mode",
			IconButton: {
				styles: {
					root: {
						mt: 10,
						letterSpacing: 250,
					},
				},
				variants: {
					custom: {
						root: {
							top: 0,
							textIndent: 1,
						},
					},
				},
			},
		};

		it("should extend styles in global", () => {
			const { container } = render(
				<IconButton variant="custom">IconButton</IconButton>,
				{
					theme: themeTest,
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend styles in local", () => {
			const { container } = render(
				<IconButton variant="custom" themeExtend={themeTest}>
					IconButton
				</IconButton>
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should render with default props correctly", () => {
			const { container } = render(<IconButton>IconButton</IconButton>, {
				theme: {
					IconButton: {
						defaultProps: {
							id: "id",
							className: "hello",
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should disable classes", () => {
			const { container } = render(<IconButton>IconButton</IconButton>, {
				theme: {
					IconButton: {
						overrideClasses: {
							root: true,
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override classes", () => {
			const { container } = render(<IconButton>IconButton</IconButton>, {
				theme: {
					IconButton: {
						overrideClasses: {
							root: "morning",
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override styles", () => {
			const { container } = render(<IconButton>IconButton</IconButton>, {
				theme: {
					IconButton: {
						overrides: {
							root: {
								mt: 50,
							},
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
