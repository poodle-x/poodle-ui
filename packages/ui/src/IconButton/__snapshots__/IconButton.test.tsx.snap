// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> ref should work 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: 0;
  padding: 8px;
  pointer-events: initial;
  font-size: 1.5rem;
  color: #282828;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: background-color 220ms ease-in,color 220ms ease-in;
  transition: background-color 220ms ease-in,color 220ms ease-in;
  border-radius: 50%;
}

.c1:hover,
.c1:focus {
  color: #282828;
  background-color: #E0E0E0;
}

.c1:disabled {
  color: #A6A6A6;
  cursor: not-allowed;
}

<button
  class="poodle-box c0 c1 c2"
>
  IconButton
</button>
`;

exports[`<IconButton /> should render Box props correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: 0;
  padding: 8px;
  pointer-events: initial;
  font-size: 1.5rem;
  color: #282828;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: background-color 220ms ease-in,color 220ms ease-in;
  transition: background-color 220ms ease-in,color 220ms ease-in;
  border-radius: 50%;
}

.c1:hover,
.c1:focus {
  color: #282828;
  background-color: #E0E0E0;
}

.c1:disabled {
  color: #A6A6A6;
  cursor: not-allowed;
}

.c2 {
  margin-top: 20px;
}

<button
  class="poodle-box c0 c1 c2"
  id="id"
>
  IconButton
</button>
`;

exports[`<IconButton /> should render IconButton correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: 0;
  padding: 8px;
  pointer-events: initial;
  font-size: 1.5rem;
  color: #282828;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: background-color 220ms ease-in,color 220ms ease-in;
  transition: background-color 220ms ease-in,color 220ms ease-in;
  border-radius: 50%;
}

.c1:hover,
.c1:focus {
  color: #282828;
  background-color: #E0E0E0;
}

.c1:disabled {
  color: #A6A6A6;
  cursor: not-allowed;
}

<button
  class="poodle-box c0 c1 c2"
>
  IconButton
</button>
`;

exports[`<IconButton /> should render shape IconButton correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: 0;
  padding: 8px;
  pointer-events: initial;
  font-size: 1.5rem;
  color: #282828;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: background-color 220ms ease-in,color 220ms ease-in;
  transition: background-color 220ms ease-in,color 220ms ease-in;
  border-radius: 4px;
}

.c1:hover,
.c1:focus {
  color: #282828;
  background-color: #E0E0E0;
}

.c1:disabled {
  color: #A6A6A6;
  cursor: not-allowed;
}

<button
  class="poodle-box c0 c1 c2"
>
  IconButton
</button>
`;

exports[`<IconButton /> theming should disable classes 1`] = `
.c0 {
  box-sizing: border-box;
}

<button
  class="poodle-box c0 c1"
>
  IconButton
</button>
`;

exports[`<IconButton /> theming should extend styles in global 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: 0;
  padding: 8px;
  pointer-events: initial;
  font-size: 1.5rem;
  color: #282828;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: background-color 220ms ease-in,color 220ms ease-in;
  transition: background-color 220ms ease-in,color 220ms ease-in;
  border-radius: 50%;
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

.c1:hover,
.c1:focus {
  color: #282828;
  background-color: #E0E0E0;
}

.c1:disabled {
  color: #A6A6A6;
  cursor: not-allowed;
}

<button
  class="poodle-box c0 c1 c2"
>
  IconButton
</button>
`;

exports[`<IconButton /> theming should extend styles in local 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: 0;
  padding: 8px;
  pointer-events: initial;
  font-size: 1.5rem;
  color: #282828;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: background-color 220ms ease-in,color 220ms ease-in;
  transition: background-color 220ms ease-in,color 220ms ease-in;
  border-radius: 50%;
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

.c1:hover,
.c1:focus {
  color: #282828;
  background-color: #E0E0E0;
}

.c1:disabled {
  color: #A6A6A6;
  cursor: not-allowed;
}

<button
  class="poodle-box c0 c1 c2"
>
  IconButton
</button>
`;

exports[`<IconButton /> theming should override classes 1`] = `
.c0 {
  box-sizing: border-box;
}

<button
  class="poodle-box c0 morning c1"
>
  IconButton
</button>
`;

exports[`<IconButton /> theming should override styles 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  margin-top: 50px;
}

<button
  class="poodle-box c0 c1 c2"
>
  IconButton
</button>
`;

exports[`<IconButton /> theming should render with default props correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border: 0;
  padding: 8px;
  pointer-events: initial;
  font-size: 1.5rem;
  color: #282828;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-transition: background-color 220ms ease-in,color 220ms ease-in;
  transition: background-color 220ms ease-in,color 220ms ease-in;
  border-radius: 50%;
}

.c1:hover,
.c1:focus {
  color: #282828;
  background-color: #E0E0E0;
}

.c1:disabled {
  color: #A6A6A6;
  cursor: not-allowed;
}

<button
  class="poodle-box c0 c1 hello c2"
  id="id"
>
  IconButton
</button>
`;
